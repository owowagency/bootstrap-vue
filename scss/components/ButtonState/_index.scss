.button-state-icon {
    vertical-align: $spinner-vertical-align;
}
