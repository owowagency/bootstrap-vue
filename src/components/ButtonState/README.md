# ButtonState

```vue
<ButtonState :error="error" :loading="loading" :success="success">
    Submit
</ButtonState>
```

## Imported Props

| Prop name      | Description                                       | Type                                             | Values | Default   |
| -------------- | ------------------------------------------------- | ------------------------------------------------ | ------ | --------- |
| errorVariant   | The variant of the component in the error state   | [Variant](../../composables/useVariant) (string) | -      | 'danger'  |
| successVariant | The variant of the component in the success state | [Variant](../../composables/useVariant) (string) | -      | 'success' |
| variant        | The variant of the component                      | [Variant](../../composables/useVariant) (string) | -      | 'primary' |

## Props

| Prop name      | Description                              | Type                   | Values | Default |
| -------------- | ---------------------------------------- | ---------------------- | ------ | ------- |
| error          | Controls the error state of the button   | boolean                | -      | false   |
| errorVariant   |                                          | variantProp('danger')  | -      |         |
| loading        | Controls the loading state of the button | boolean                | -      | false   |
| success        | Controls the success state of the button | boolean                | -      | false   |
| successVariant |                                          | variantProp('success') | -      |         |
| variant        |                                          | variantProp()          | -      |         |

## Slots

| Name    | Description                        | Bindings |
| ------- | ---------------------------------- | -------- |
| default | Displays content inside the button |          |
