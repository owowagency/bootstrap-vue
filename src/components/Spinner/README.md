# Spinner

```vue
<Spinner size="lg" />
```

## Imported Props

| Prop name | Description                  | Type                                             | Values | Default   |
| --------- | ---------------------------- | ------------------------------------------------ | ------ | --------- |
| size      | The size of the component    | [Size](../../composables/useSize) (string)       | -      | 'md'      |
| variant   | The variant of the component | [Variant](../../composables/useVariant) (string) | -      | 'primary' |

## Props

| Prop name | Description             | Type | Values             | Default  |
| --------- | ----------------------- | ---- | ------------------ | -------- |
| type      | The type of the spinner | Type | `'border'\|'grow'` | 'border' |
