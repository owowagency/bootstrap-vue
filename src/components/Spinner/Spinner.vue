<template>
    <div :class="classes" />
</template>

<script lang="ts">
import {PropType, computed, toRef} from 'vue';
import useSize, {sizeProps} from '@/composables/useSize';
import useVariant, {variantProps} from '@/composables/useVariant';
import useClasses from '@/composables/useClasses';
import useStringTemplate from '@/composables/useStringTemplate';

export const types = ['border', 'grow'] as const;

type Type = typeof types[number];
</script>

<script lang="ts" setup>
const props = defineProps({
    ...sizeProps,
    /**
     * The type of the spinner
     * @values 'border'|'grow'
     */
    type: {
        type: String as PropType<Type>,
        default: 'border',
        validator: (t: Type) => types.includes(t),
    },
    ...variantProps,
});

const spinnerClass = useStringTemplate('spinner-{0}', toRef(props, 'type')).templatedString;

const {classes} = useClasses(computed(() => [
    useSize(props.size, `${spinnerClass.value}-{0}`).sizeClass.value,
    spinnerClass.value,
    useVariant(props.variant, 'text-{0}').variantClass.value,
]));
</script>

<docs>
```vue
<Spinner size="lg" />
```

## Imported Props

| Prop name | Description                  | Type                                             | Values | Default   |
| --------- | ---------------------------- | ------------------------------------------------ | ------ | --------- |
| size      | The size of the component    | [Size](../../composables/useSize) (string)       | -      | 'md'      |
| variant   | The variant of the component | [Variant](../../composables/useVariant) (string) | -      | 'primary' |
</docs>
