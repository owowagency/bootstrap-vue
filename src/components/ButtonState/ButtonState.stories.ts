import ButtonState from '.';

export default {
    title: 'Components/Button state',
    component: ButtonState,
    argTypes: {},
};

const template = (args) => ({
    components: {ButtonState},
    setup: () => ({args}),
    template: '<ButtonState v-bind="args">Button state</ButtonState>',
});

export const Loading = template.bind({});

Loading.args = {loading: true};

export const Error = template.bind({});

Error.args = {error: true};

export const Success = template.bind({});

Success.args = {success: true};
