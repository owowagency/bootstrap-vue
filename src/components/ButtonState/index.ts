import ButtonState from './ButtonState.vue';

export default ButtonState;
