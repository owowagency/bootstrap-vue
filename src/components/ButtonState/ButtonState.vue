<template>
    <Button
        :disabled="disabled"
        :variant="variant"
    >
        <span
            v-if="error || loading || success"
            class="me-2"
        >
            <Spinner
                v-if="loading"
                size="sm"
                variant=""
            />

            <svg
                v-if="success"
                class="button-state-icon"
                data-test="success"
                fill="currentColor"
                height="1em"
                viewBox="3 3 10 10"
                xmlns="http://www.w3.org/2000/svg"
                width="1em"
            >
                <path d="M10.97 4.97a.75.75 0 0 1 1.07 1.05l-3.99 4.99a.75.75 0 0 1-1.08.02L4.324 8.384a.75.75 0 1 1 1.06-1.06l2.094 2.093 3.473-4.425a.267.267 0 0 1 .02-.022z" />
            </svg>

            <svg
                v-if="error"
                class="button-state-icon"
                data-test="error"
                fill="currentColor"
                height="1em"
                viewBox="0 0 16 16"
                xmlns="http://www.w3.org/2000/svg"
                width="1em"
            >
                <path d="M8 15A7 7 0 1 1 8 1a7 7 0 0 1 0 14zm0 1A8 8 0 1 0 8 0a8 8 0 0 0 0 16z" />
                <path d="M7.002 11a1 1 0 1 1 2 0 1 1 0 0 1-2 0zM7.1 4.995a.905.905 0 1 1 1.8 0l-.35 3.507a.552.552 0 0 1-1.1 0L7.1 4.995z" />
            </svg>
        </span>

        <!-- @slot Displays content inside the button -->
        <slot />
    </Button>
</template>

<script lang="ts">
import {Variant, variantProp} from '@/composables/useVariant';
import Button from '@/components/Button';
import Spinner from '@/components/Spinner';
import {computed} from 'vue';
</script>

<script lang="ts" setup>
const props = defineProps({
    /**
     * Controls the error state of the button
     */
    error: {
        type: Boolean,
        default: false,
    },
    errorVariant: variantProp('danger'),
    /**
     * Controls the loading state of the button
     */
    loading: {
        type: Boolean,
        default: false,
    },
    /**
     * Controls the success state of the button
     */
    success: {
        type: Boolean,
        default: false,
    },
    successVariant: variantProp('success'),
    variant: variantProp(),
});

const disabled = computed(() => props.loading);

const variant = computed<Variant>(() => {
    switch (true) {
        case props.success:
            return props.successVariant;

        case props.error:
            return props.errorVariant;

        default:
            return props.variant;
    }
});
</script>

<docs>
```vue
<ButtonState
    :error="error"
    :loading="loading"
    :success="success"
>
    Submit
</ButtonState>
```

## Imported Props

| Prop name      | Description                                       | Type                                             | Values | Default   |
| -------------- | ------------------------------------------------- | ------------------------------------------------ | ------ | --------- |
| errorVariant   | The variant of the component in the error state   | [Variant](../../composables/useVariant) (string) | -      | 'danger'  |
| successVariant | The variant of the component in the success state | [Variant](../../composables/useVariant) (string) | -      | 'success' |
| variant        | The variant of the component                      | [Variant](../../composables/useVariant) (string) | -      | 'primary' |
</docs>
