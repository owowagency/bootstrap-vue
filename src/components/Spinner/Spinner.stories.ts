import Spinner from '.';
import {sizes} from '@/composables/useSize';
import {types} from './Spinner.vue';
import {variants} from '@/composables/useVariant';

export default {
    title: 'Components/Spinner',
    component: Spinner,
    argTypes: {
        type: {
            control: {type: 'select'},
            options: types,
        },
        size: {
            control: {type: 'select'},
            options: sizes,
        },
        variant: {
            control: {type: 'select'},
            options: variants,
        },
    },
};

const template = (args) => ({
    components: {Spinner},
    setup: () => ({args}),
    template: '<Spinner v-bind="args" />',
});

export const Default = template.bind({});

export const Sizing = (args) => ({
    components: {Spinner},
    setup: () => ({args, sizes}),
    template: '<Spinner v-for="s of sizes" class="me-1" :size="s" v-bind="args" />',
});

export const Type = (args) => ({
    components: {Spinner},
    setup: () => ({args, types}),
    template: '<Spinner v-for="t of types" class="me-1" :type="t" v-bind="args" />',
});

export const Variants = (args) => ({
    components: {Spinner},
    setup: () => ({args, variants}),
    template: '<Spinner v-for="v of variants" class="me-1" :variant="v" v-bind="args" />',
});
