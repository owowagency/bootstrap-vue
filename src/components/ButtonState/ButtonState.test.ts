import ButtonState from '.';
import Spinner from '@/components/Spinner';
import {shallowMount} from '@vue/test-utils';

describe('template', () => {
    componentRenderTest(ButtonState);

    componentSlotRenderTest(ButtonState);

    it('renders spinner when loading', () => {
        const wrapper = shallowMount(ButtonState, {
            props: {loading: true},
            global: {renderStubDefaultSlot: true},
        });

        expect(wrapper.findComponent(Spinner).exists()).toBe(true);
    });

    it('renders success when success', () => {
        const wrapper = shallowMount(ButtonState, {
            props: {success: true},
            global: {renderStubDefaultSlot: true},
        });

        expect(wrapper.find('[data-test="success"]').exists()).toBe(true);
    });

    it('renders error when error', () => {
        const wrapper = shallowMount(ButtonState, {
            props: {error: true},
            global: {renderStubDefaultSlot: true},
        });

        expect(wrapper.find('[data-test="error"]').exists()).toBe(true);
    });
});

describe('disabled', () => {
    it('is false when not disabled', () => {
        const wrapper = shallowMount(ButtonState, {props: {disabled: false}});

        expect(wrapper.vm.disabled).toBe(false);
    });

    it('is true when disabled', () => {
        const wrapper = shallowMount(ButtonState, {props: {disabled: true}});

        expect(wrapper.vm.disabled).toBe(true);
    });

    it('is true when disabled and not loading', () => {
        const wrapper = shallowMount(ButtonState, {props: {disabled: true, loading: false}});

        expect(wrapper.vm.disabled).toBe(true);
    });

    it('is false when not loading', () => {
        const wrapper = shallowMount(ButtonState, {props: {loading: false}});

        expect(wrapper.vm.disabled).toBe(false);
    });

    it('is true when loading', () => {
        const wrapper = shallowMount(ButtonState, {props: {loading: true}});

        expect(wrapper.vm.disabled).toBe(true);
    });
});

describe('variant', () => {
    it('is variant prop by default', () => {
        const wrapper = shallowMount(ButtonState, {props: {variant: 'info'}});

        expect(wrapper.vm.variant).toBe('info');
    });

    it('is successVariant prop when success', () => {
        const wrapper = shallowMount(ButtonState, {
            props: {
                success: true,
                successVariant: 'info',
            },
        });

        expect(wrapper.vm.variant).toBe('info');
    });

    it('is errorVariant prop when error', () => {
        const wrapper = shallowMount(ButtonState, {
            props: {
                error: true,
                errorVariant: 'info',
            },
        });

        expect(wrapper.vm.variant).toBe('info');
    });
});
