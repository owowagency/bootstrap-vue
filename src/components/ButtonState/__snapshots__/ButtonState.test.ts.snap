// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template renders default 1`] = `"<button-stub outline=\\"false\\" disabled=\\"false\\" tag=\\"button\\" size=\\"md\\" variant=\\"primary\\"></button-stub>"`;
